File: Easy/290_Word_Pattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        dic = {}
        words = s.split()
        if len(words) != len(pattern):
            return False
        for index, word in enumerate(words):
            if(pattern[index] not in dic.keys()):
                if(word in dic.values()):
                    return False
                else:
                    dic[pattern[index]] = word
            else:
                if(dic[pattern[index]] != word):
                    return False

        return True 

sol = Solution()
pattern = "abba"
s = "dog dog dog dog"
print(sol.wordPattern(pattern, s))

File: EasyJS/290_Word_Pattern.js
var wordPattern = function (pattern, s) {
  let pToW = new Map();
  let wToP = new Map();
  let words = s.split(" ");

  if (pattern.length !== words.length) return false;

  for (let i = 0; i < pattern.length; i++) {
    let p = pattern[i];
    let w = words[i];

    if (
      (pToW.has(p) && pToW.get(p) !== w) ||
      (wToP.has(w) && wToP.get(w) !== p)
    ) {
      return false;
    }

    pToW.set(p, w);
    wToP.set(w, p);
  }

  return true;
};
